### ISSUE TITLE
Lack of warning for getting pointer on a temporary CString declared in array

### ISSUE THREAD

**Issue by PeakKS** [rust-lang/rust#101398]
Given the following code: 

```rust
let instance_layer_names: Vec<_> = [
	CString::new("VK_LAYER_KHRONOS_validation").unwrap()
].iter().map(|l| l.as_ptr()).collect();
```

There is no error / warning reported, but the string is invalid. If I was to try to pass this variable to the InstanceCreateInfoBuilder in ash for example I would get an `ERROR_LAYER_NOT_PRESENT`.

I was able to fix my code by splitting that expression as shown:
```rust
let instance_layers = [
    CString::new("VK_LAYER_KHRONOS_validation").unwrap()
];

let instance_layer_names: Vec<_> = instance_layers.iter().map(|l| l.as_ptr()).collect(); 
```
While testing before writing this error report I noticed that if I have that new code alongside my original code the original code will actually work given that they both have the same string value.

Ideally the assignment I was trying to do originally should give a warning/error, I assume the warning would be

```
warning: getting the inner pointer of a temporary `CString`
```

since that seems to be the cause of the issue.
